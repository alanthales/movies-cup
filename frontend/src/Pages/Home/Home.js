import React from "react";
import Snackbar from "@material-ui/core/Snackbar";
import Layout from "../../Components/Layout/Layout";
import { MoviesList, MoviesCount, MAX_LIST } from "../../Components/Movies";
import withFetch from "../../Components/Fetch/withFetch";

export class Home extends React.Component {
  state = {
    moviesSelecteds: [],
    selectedsCount: 0,
    notes: null
  }

  handleSelect(event, checked) {
    let movie = this.props.data.find(x => x.id === event.target.value);

    this.setState((state) => {
      if (checked) {
        return {
          selectedsCount: state.selectedsCount + 1,
          moviesSelecteds: [...state.moviesSelecteds, movie]
        };
      }

      return {
        selectedsCount: state.selectedsCount - 1,
        moviesSelecteds: state.moviesSelecteds.filter(x => x.id !== movie.id)
      };
    });
  }

  handleGenerate() {
    if (this.props.hasError) return;
    
    if (this.state.selectedsCount !== MAX_LIST) {
      this.setState({ notes: "Você deve selecionar 8 filmes" });
      return;
    }

    const { moviesSelecteds: movies } = this.state;
    this.props.history.push("/classification", { movies });
  }

  render() {
    const { selectedsCount, notes } = this.state,
      { data, isLoading, hasError } = this.props;

    return (
      <div>
        <Layout
          title="Fase de Seleção"
          subtitle="Selecione 8 filmes que você deseja que entre na competição e depois pressione o botão Gerar Meu Campeonato para prosseguir."
        >
          <MoviesCount value={selectedsCount} onGenerate={this.handleGenerate.bind(this)} />

          {isLoading && <p className="text-center">Carregando...</p>}

          {hasError && <p className="text-center">Opps! Algo deu errado, tente recarregar a página.</p>}

          {!isLoading && !hasError && (
            <MoviesList movies={data} onSelect={this.handleSelect.bind(this)} />
          )}
        </Layout>

        <Snackbar
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'right',
          }}
          open={!!notes}
          autoHideDuration={5000}
          onClose={() => this.setState({ notes: null })}
          message={notes}
        />
      </div>
    );
  }
}

const fetchOptions = { url: "/api/movies" };
const hocHome = withFetch(fetchOptions)(Home);

export default hocHome;