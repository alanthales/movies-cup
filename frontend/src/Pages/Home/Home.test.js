import React from "react";
import renderer from "react-test-renderer";
import { Home } from "./Home";

test("render Home with isLoading prop", () => {
  const component = renderer.create(
    <Home isLoading={true} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("render Home with hasError prop", () => {
  const component = renderer.create(
    <Home isLoading={false} hasError={true} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("render Home with data prop", () => {
  const data = [{ titulo: "Teste", ano: 2019 }];
  const component = renderer.create(
    <Home isLoading={false} hasError={false} data={data} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});