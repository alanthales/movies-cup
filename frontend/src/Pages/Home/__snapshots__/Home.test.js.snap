// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Home with data prop 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiBox-root MuiBox-root-144"
      >
        <div
          className="MuiGrid-root header-component MuiGrid-container MuiGrid-justify-xs-center"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
          >
            <h4>
              CAMPEONATO DE FILMES
            </h4>
            <h1>
              Fase de Seleção
            </h1>
            <hr
              style={
                Object {
                  "width": "50px",
                }
              }
            />
            <p>
              Selecione 8 filmes que você deseja que entre na competição e depois pressione o botão Gerar Meu Campeonato para prosseguir.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiGrid-root movies-component-count MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <h4>
            Selecionados
          </h4>
          <p>
            0 de 8
          </p>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Gerar Meu Campeonato
            </span>
          </button>
        </div>
      </div>
      <div
        className="MuiGrid-root movies-component-list MuiGrid-container MuiGrid-justify-xs-space-between"
      >
        <div
          className="MuiGrid-root movies-component-list-item MuiGrid-item"
        >
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
          >
            <div
              className="MuiGrid-root MuiGrid-item"
            >
              <span
                aria-disabled={false}
                className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-151 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={null}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <input
                    className="PrivateSwitchBase-input-154"
                    data-indeterminate={false}
                    onChange={[Function]}
                    type="checkbox"
                  />
                </span>
              </span>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item"
            >
              <h5>
                Teste
              </h5>
              <p>
                2019
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render Home with hasError prop 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiBox-root MuiBox-root-143"
      >
        <div
          className="MuiGrid-root header-component MuiGrid-container MuiGrid-justify-xs-center"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
          >
            <h4>
              CAMPEONATO DE FILMES
            </h4>
            <h1>
              Fase de Seleção
            </h1>
            <hr
              style={
                Object {
                  "width": "50px",
                }
              }
            />
            <p>
              Selecione 8 filmes que você deseja que entre na competição e depois pressione o botão Gerar Meu Campeonato para prosseguir.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiGrid-root movies-component-count MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <h4>
            Selecionados
          </h4>
          <p>
            0 de 8
          </p>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Gerar Meu Campeonato
            </span>
          </button>
        </div>
      </div>
      <p
        className="text-center"
      >
        Opps! Algo deu errado, tente recarregar a página.
      </p>
    </div>
  </div>
</div>
`;

exports[`render Home with isLoading prop 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiBox-root MuiBox-root-105"
      >
        <div
          className="MuiGrid-root header-component MuiGrid-container MuiGrid-justify-xs-center"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
          >
            <h4>
              CAMPEONATO DE FILMES
            </h4>
            <h1>
              Fase de Seleção
            </h1>
            <hr
              style={
                Object {
                  "width": "50px",
                }
              }
            />
            <p>
              Selecione 8 filmes que você deseja que entre na competição e depois pressione o botão Gerar Meu Campeonato para prosseguir.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiGrid-root movies-component-count MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <h4>
            Selecionados
          </h4>
          <p>
            0 de 8
          </p>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Gerar Meu Campeonato
            </span>
          </button>
        </div>
      </div>
      <p
        className="text-center"
      >
        Carregando...
      </p>
    </div>
  </div>
</div>
`;
